
## Overview
Code for analyses presented in Single-cell multimodal glioma analyses reveal epigenetic regulators of cellular plasticity and environmental stress response. [Paper link](https://www.nature.com/articles/s41588-021-00926-8).

## Data availability
Processed data is available at: https://www.synapse.org/singlecellglioma

## Abstract
Glioma intratumoral heterogeneity enables adaptation to challenging microenvironments and contributes to therapeutic resistance. We integrated 914 single-cell DNA methylomes, 55,284 single-cell transcriptomes, and bulk multi-omic profiles across 11 adult IDH-mutant or IDH-wild-type gliomas to delineate sources of intratumoral heterogeneity. Local DNA methylation disorder was associated with cell-to-cell DNA methylation differences, elevated in more aggressive tumors, linked with transcriptional disruption, and associated with environmental stress response. We demonstrated that in vitro, cells under hypoxia and irradiation stress increase local DNA methylation disorder and shift cell states. We identified a positive association between genetic and epigenetic instability that was supported in cohorts of bulk longitudinally collected DNA methylation data. Increased DNA methylation disorder associated with accelerated disease progression, and recurrently selected DNA methylation changes were enriched for environmental stress response pathways. Our work identifies an epigenetically facilitated adaptive stress response process and highlights the importance of epigenetic heterogeneity in shaping therapeutic outcomes.


## Contact
Lead contact for data and materials availability: [Roel Verhaak](https://verhaaklab.com/contact/)

Lead analysts: Kevin J. Anderson and Kevin C. Johnson
